### PR TITLE
Add static build support

### LINKED ISSUE #220: Add static build
As with xeus, it would be nice to have the possibility to link to a static target (e.g. `xwidget-static`) given that the library is not available in many distros

### PR DESCRIPTION
**Pull request by rapgenic**
This PR adds the static library compilation and fixes #220.

I have marked it as a draft because as of now I have not been able to test if the shared library compilation still works (I need the static libraries because I do not have all the shared dependencies installed)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -159,75 +165,117 @@ set(XWIDGETS_SOURCES
...
+    target_include_directories(${target_name} PUBLIC $<BUILD_INTERFACE:${XWIDGETS_INCLUDE_DIR}>
+                                               $<INSTALL_INTERFACE:include>)
 
-include(CheckCXXCompilerFlag)
-string(TOUPPER "${CMAKE_BUILD_TYPE}" U_CMAKE_BUILD_TYPE)
-OPTION(DISABLE_ARCH_NATIVE "disable -march=native flag" OFF)
+    target_link_libraries(${target_name}
+                          PUBLIC xtl
+                          PUBLIC ${XEUS_TARGET_NAME}
+                          PUBLIC xproperty)
```
Comment (rapgenic) [inline CMakeLists.txt:196]: This was necessary to allow builds with in-tree xproperty (obtained with cmake's FetchContent). Without it all includes to xproperty files are not found

**Comment by SylvainCorlay** [comment]
Thanks @rapgenic for tackling this!

I think it makes a lot of sense to provide a static build of xwidgets, for many reasons. I just looked it up and found your xeus-octave project, which I think is a great addition. Kudos on this!

**Comment by rapgenic** [comment]
I can see that all the tests have passed, can this be considered enough to prove that there is no breaking change? (I am asking that because I'm having a few difficulties to build all the dependencies on my system)

Than you for your nice words on my project! I must say however that is still very much a proof of concept and I'm experimenting with many things at the moment, integration the xwidgets library is one of the next things I'll be checking out

File: CMakeLists.txt [hunk 2]
```diff
@@ -159,75 +166,116 @@ set(XWIDGETS_SOURCES
...
+            if (HAS_LTO_FLAG)
+                target_compile_options(${target_name} PUBLIC -flto)
+            endif()
+
+            # Intel equivalent to LTO is called IPO
+            if (CMAKE_CXX_COMPILER_ID MATCHES "Intel")
+                CHECK_CXX_COMPILER_FLAG("-ipo" HAS_IPO_FLAG)
+                if (HAS_IPO_FLAG)
+                    target_compile_options(${target_name} PUBLIC -ipo)
+                endif()
```
Comment (SylvainCorlay) [inline CMakeLists.txt:239]: I think we should completely remote the section about LTO/IPO from the cmake.

File: CMakeLists.txt (hunk 1 above)
Comment (SylvainCorlay) [reply CMakeLists.txt:197]: :+1:

File: CMakeLists.txt (hunk 2 above)
Comment (rapgenic) [reply CMakeLists.txt:239]: Ok! (I had kept it as I did not know what was for)

**Comment by SylvainCorlay** [comment]
> I can see that all the tests have passed, can this be considered enough to prove that there is no breaking change? (I am asking that because I'm having a few difficulties to build all the dependencies on my system)

I think that it looks good. For your own development workflow, I recommend using conda even for octave etc in your development workflow. Octave is already packaged on conda-forge as well as the xeus stack.

> Than you for your nice words on my project! I must say however that is still very much a proof of concept and I'm experimenting with many things at the moment, integration with the xwidgets library is one of the next things I'll be checking out!

Making a xeus-octave is really something that was on our radar for some time. (I did some nights-and-weekend work towards scilab, but Scilab resisted a bit).

**Comment by SylvainCorlay** [comment]
cc @JohanMabille if you want to have a look at the changes.

**Comment by rapgenic** [comment]
> I recommend using conda even for octave etc in your development workflow. Octave is already packaged on conda-forge as well as the xeus stack.

Thank you for the suggestion, I have no experience with conda, but I'll give it a look. At the moment however I have no problem building with static linkage (the only problem was installing all the dependencies to test the non-static build) and I think it is enough for the development purposes, as I don't have to manage any dependency at the expense of having a bigger binary. Once the project is more mature I'll surely look at integrating it, as it seems to me that conda has become the main way of distributing software in this ecosystem.

> I think that it looks good

I'll remove the draft marker, then. 

> Making a xeus-octave is really something that was on our radar for some time. (I did some nights-and-weekend work towards scilab, but Scilab resisted a bit).

Octave, too, is quite convoluted.. I remember however that there's a patch done a few years ago by some GSOC student that tried to implement a basic kernel if you're interested

**Comment by SylvainCorlay** [comment]
I will merge today. Leaving it open for a bit to give the occasion to @JohanMabille to comment!

**Comment by SylvainCorlay** [comment]
Thanks!

**Pull request merged**